File: app/src/main/java/com/example/opengl/Resource.java
package com.example.opengl;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.opengl.devourer.DevourerNode;
import com.example.opengl.devourer.DevourerStructure;
import com.example.opengl.tiles.MineralType;

public class Resource {
    public DevourerStructure devourerStructure;
    public MineralType mineralType;
    public Position startPosition;
    public Position nextPosition;
    public PositionInd startPositionInd;
    public PositionInd nextPositionInd;
    public float distance; // 0...1
    public boolean isMoving;
    public boolean isDelivered;

    public Resource(MineralType mineralType) {
        this.mineralType = mineralType;
        this.distance = 0;
        this.isMoving = false;
        this.isDelivered = false;
    }

    public void startResource(@NonNull PositionInd startPositionInd, Position startPosition, DevourerStructure devourerStructure) {
        this.devourerStructure = devourerStructure;
        this.startPosition = startPosition;
        this.nextPosition = null;
        this.startPositionInd = startPositionInd;
        this.nextPositionInd = calculateNextPositionInd(startPositionInd);
        this.distance = 0;
        this.isMoving = true;
        this.isDelivered = false;
    }

    public void move(float delta) {
        DevourerNode startDevourerNode = startPositionInd != null ? devourerStructure.getDevourerNode(startPositionInd.x, startPositionInd.y) : null;
        DevourerNode nextDevourerNode = nextPositionInd != null ? devourerStructure.getDevourerNode(nextPositionInd.x, nextPositionInd.y) : null;

        if ((startDevourerNode == null) && ((distance + delta) <= 0.5f)) { // in deleted start tile
            // stopped in startPosition
            isMoving = false;
            return;
        }
        if (nextDevourerNode == null) {
            if ((distance + delta) > 0.5f) {
                isMoving = false;
                // stopped in nextPosition
                return;
            } else {
                // stopped in startPosition;
                isMoving = false;
                return;
            }
        }
        float distanceOverflow = delta + distance - 1f;
        if (distanceOverflow > 0) { // move to the next tile
            if (nextDevourerNode.dist == 0) {
                distance = 1f;
                isMoving = false;
                isDelivered = true;
                return;
            }
            startPositionInd = nextPositionInd;
            nextPositionInd = calculateNextPositionInd(startPositionInd);
            distance = distanceOverflow;
            startPosition = nextPosition;
            nextPosition = null;
        } else {
            distance += delta;
        }
    }

    @Nullable
    public PositionInd calculateNextPositionInd(@NonNull PositionInd positionInd) {
        PositionInd result = null;
        DevourerNode devourerNode = devourerStructure.getDevourerNode(positionInd.x, positionInd.y);
        if (devourerNode != null) {
            int dist = devourerNode.dist - 1;
            for (DevourerNode neighbor : devourerNode.getNeighbors()) {
                if (neighbor.dist == dist) {
                    result = new PositionInd();
                    result.x = neighbor.x;
                    result.y = neighbor.y;
                    break;
                }
            }
        }
        return result;
    }

}

File: app/src/main/java/com/example/opengl/Resource.kt
package com.example.opengl

import com.example.opengl.devourer.DevourerStructure
import com.example.opengl.tiles.MineralType

class Resource(var mineralType: MineralType) {
    var devourerStructure: DevourerStructure? = null
    var startPosition: Position? = null
    var nextPosition: Position? = null
    var startPositionInd: PositionInd? = null
    var nextPositionInd: PositionInd? = null
    var distance // 0...1
            = 0f
    var isMoving = false
    var isDelivered = false

    fun startResource(
        startPositionInd: PositionInd,
        startPosition: Position?,
        devourerStructure: DevourerStructure?
    ) {
        this.devourerStructure = devourerStructure
        this.startPosition = startPosition
        nextPosition = null
        this.startPositionInd = startPositionInd
        nextPositionInd = calculateNextPositionInd(startPositionInd)
        distance = 0f
        isMoving = true
        isDelivered = false
    }

    fun move(delta: Float) {
        val startDevourerNode = if (startPositionInd != null) devourerStructure!!.getDevourerNode(
            startPositionInd!!.x, startPositionInd!!.y
        ) else null
        val nextDevourerNode = if (nextPositionInd != null) devourerStructure!!.getDevourerNode(
            nextPositionInd!!.x, nextPositionInd!!.y
        ) else null
        if (startDevourerNode == null && distance + delta <= 0.5f) { // in deleted start tile
            // stopped in startPosition
            isMoving = false
            return
        }
        if (nextDevourerNode == null) {
            if (distance + delta > 0.5f) {
                isMoving = false
                // stopped in nextPosition
                return
            } else {
                // stopped in startPosition;
                isMoving = false
                return
            }
        }
        val distanceOverflow = delta + distance - 1f
        if (distanceOverflow > 0) { // move to the next tile
            if (nextDevourerNode.dist == 0) { // Delivered to main devourer
                distance = 1f
                isMoving = false
                isDelivered = true
                return
            }
            startPositionInd = nextPositionInd
            nextPositionInd = calculateNextPositionInd(startPositionInd!!)
            distance = distanceOverflow
            startPosition = nextPosition
            nextPosition = null
        } else {
            distance += delta
        }
    }

    fun tryMove(delta: Float) {
        if (startPositionInd  != null) {
            nextPositionInd = calculateNextPositionInd(startPositionInd!!)
            if (nextPositionInd != null) {
                move(delta)
            }
        }
    }

    private fun calculateNextPositionInd(positionInd: PositionInd): PositionInd? {
        var result: PositionInd? = null
        //val devourerNode = devourerStructure!!.getDevourerNode(positionInd.x, positionInd.y)
        val devourerNode = devourerStructure?.getDevourerNode(positionInd.x, positionInd.y)
        if (devourerNode != null) {
            val dist = devourerNode.dist - 1
            for (neighbor in devourerNode.getNeighbors()) {
                if (neighbor.dist == dist) {
                    result = PositionInd(neighbor.x, neighbor.y)
                    break
                }
            }
        }
        return result
    }
}